flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/PhysicalSlotRequestBulkTracker.java

```diff
@@ -0,0 +1,164 @@
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Tracks physical slot request bulks. Once a bulk is registered, a fulfill-ability check for it will be scheduled.
+ */
+class PhysicalSlotRequestBulkTracker {
+
+	private final Supplier<Set<SlotInfo>> slotsRetriever;
+
+	private final Clock clock;
+
+	private final Set<PhysicalSlotRequestBulk> slotRequestBulks;
+
+	PhysicalSlotRequestBulkTracker(final Supplier<Set<SlotInfo>> slotsRetriever, final Clock clock) {
+		this.slotsRetriever = checkNotNull(slotsRetriever);
+		this.clock = checkNotNull(clock);
+		this.slotRequestBulks = Collections.newSetFromMap(new IdentityHashMap<>());
+	}
+
+	void track(final PhysicalSlotRequestBulk bulk) {
+		slotRequestBulks.add(bulk);
+
+		bulk.markUnfulfillable(clock.relativeTimeMillis());
+	}
+
+	void untrack(final PhysicalSlotRequestBulk bulk) {
+		slotRequestBulks.remove(bulk);
+	}
+
+	@VisibleForTesting
+	boolean isTracked(final PhysicalSlotRequestBulk bulk) {
+		return slotRequestBulks.contains(bulk);
+	}
+
+	/**
+	 * Check the slot request bulk and timeout its requests if it has been unfilfillable for too long.
+	 * @param slotRequestBulk bulk of slot requests
+	 * @param slotRequestTimeout indicates how long a pending request can be unfilfillable
+	 * @return result of the check, indicating the bulk is fulfilled, still pending, or timed out
+	 */
+	TimeoutCheckResult checkPhysicalSlotRequestBulkTimeout(
+			final PhysicalSlotRequestBulk slotRequestBulk,
+			final Time slotRequestTimeout) {
+
+		checkState(isTracked(slotRequestBulk));
+
+		if (slotRequestBulk.getPendingRequests().isEmpty()) {
+			return TimeoutCheckResult.FULFILLED;
+		}
+
+		final boolean fulfillable = isSlotRequestBulkFulfillable(slotRequestBulk, slotsRetriever);
+		if (fulfillable) {
+			slotRequestBulk.markFulfillable();
+		} else {
+			final long currentTimestamp = clock.relativeTimeMillis();
+
+			slotRequestBulk.markUnfulfillable(currentTimestamp);
+
+			final long unfulfillableSince = slotRequestBulk.getUnfulfillableSince();
+			if (unfulfillableSince + slotRequestTimeout.toMilliseconds() <= currentTimestamp) {
+				return TimeoutCheckResult.TIMEOUT;
+			}
+		}
+
+		return TimeoutCheckResult.PENDING;
+	}
+
+	/**
+	 * Returns whether the given bulk of slot requests are possible to be fulfilled at the same time
+	 * with all the reusable slots in the slot pool. A reusable slot means the slot is available or
+	 * will not be occupied indefinitely.
+	 *
+	 * @param slotRequestBulk bulk of slot requests to check
+	 * @param slotsRetriever supplies slots to be used for the fulfill-ability check
+	 * @return true if the slot requests are possible to be fulfilled, otherwise false
+	 */
+	@VisibleForTesting
+	static boolean isSlotRequestBulkFulfillable(
```

Comment:
Why do we need to expose `isSlotRequestBulkFulfillable`?
Could we not test the same with `checkPhysicalSlotRequestBulkTimeout` and e.g. `PhysicalSlotRequestBulk.isFulfillable`?

Reply:
Yes it's possible to test `isSlotRequestBulkFulfillable` with `checkPhysicalSlotRequestBulkTimeout` but it will complicate the tests for understanding and maintenance.
`isSlotRequestBulkFulfillable` is a static utility method so I think it's fine to test it separately.